File: backend/finance/descriptor.go
package finance

import (
	"strings"
	"unicode/utf8"
)

// suggestedPatternWords is how many leading significant words of a descriptor
// a learned rule keys on by default ("cruz verde", "entel pcs").
const suggestedPatternWords = 2

// normalizeDescriptor reduces a bank descriptor to its stable words: lowercase,
// without tokens carrying digits (store or terminal ids such as "L9093") or
// single letters (the trailing channel code), whitespace collapsed.
// "CRUZ VERDE L9093 CHILLAN  C" → "cruz verde chillan".
func normalizeDescriptor(s string) string {
	words := make([]string, 0, 8)
	for w := range strings.FieldsSeq(strings.ToLower(s)) {
		if utf8.RuneCountInString(w) < 2 || strings.ContainsAny(w, "0123456789") {
			continue
		}
		words = append(words, w)
	}
	return strings.Join(words, " ")
}

// suggestPattern proposes the rule pattern for a descriptor: its first
// significant words, so the rule also covers other branches of the merchant.
func suggestPattern(description string) string {
	words := strings.Fields(normalizeDescriptor(description))
	return strings.Join(words[:min(len(words), suggestedPatternWords)], " ")
}

// ruleFor returns the rule whose pattern prefixes the descriptor at a word
// boundary; the longest (most specific) pattern wins, ties keep the first.
func ruleFor(rules []MerchantRule, description string) (MerchantRule, bool) {
	norm := normalizeDescriptor(description)
	var best MerchantRule
	found := false
	for _, r := range rules {
		if norm != r.Pattern && !strings.HasPrefix(norm, r.Pattern+" ") {
			continue
		}
		// Counted in runes, like the TS engine's [...pattern].length.
		if !found || utf8.RuneCountInString(r.Pattern) > utf8.RuneCountInString(best.Pattern) {
			best, found = r, true
		}
	}
	return best, found
}

File: frontend/src/engine/finance/descriptor.ts
// Mirror of backend/finance/descriptor.go: how bank descriptors are normalized,
// which rule pattern a descriptor suggests, and which learned rule applies.
import type { MerchantRule } from '@/services/contract'

// suggestedPatternWords is how many leading significant words of a descriptor a
// learned rule keys on by default ("cruz verde", "entel pcs").
const suggestedPatternWords = 2

// normalizeDescriptor reduces a bank descriptor to its stable words: lowercase,
// without tokens carrying digits (store or terminal ids such as "L9093") or
// single letters (the trailing channel code), whitespace collapsed.
// "CRUZ VERDE L9093 CHILLAN  C" → "cruz verde chillan".
export function normalizeDescriptor(s: string): string {
  return s
    .toLowerCase()
    .split(/\s+/)
    .filter((w) => [...w].length >= 2 && !/[0-9]/.test(w))
    .join(' ')
}

// suggestPattern proposes the rule pattern for a descriptor: its first
// significant words, so the rule also covers other branches of the merchant.
export function suggestPattern(description: string): string {
  const norm = normalizeDescriptor(description)
  return norm === '' ? '' : norm.split(' ').slice(0, suggestedPatternWords).join(' ')
}

// ruleFor returns the rule whose pattern prefixes the descriptor at a word
// boundary; the longest (most specific) pattern wins, ties keep the first.
export function ruleFor(rules: MerchantRule[], description: string): MerchantRule | null {
  const norm = normalizeDescriptor(description)
  let best: MerchantRule | null = null
  for (const r of rules) {
    if (norm !== r.pattern && !norm.startsWith(r.pattern + ' ')) continue
    // Counted in code points, like the Go side's utf8.RuneCountInString.
    if (!best || [...r.pattern].length > [...best.pattern].length) best = r
  }
  return best
}
